### ISSUE TITLE
wslbridge2 in WSL1 breaks connection when linked with glibc statically

### ISSUE THREAD

**Issue by Biswa96** [Biswa96/wslbridge2#15]
* **Problem:** wslbridge2 in WSL1 breaks connection when linked with glibc
statically. This is not issue for hvpty in WSL2.

* **Experiments:** 

  - This issue does not occur with _dynamic linking_. Also static linking with
  _musl libc_ does not produce any error. Hence the v0.4 released binaries are
  compiled with musl libc statically.

  - This issue is inherited from the base wslbridge repository by @rprichard.
  To reproduce this issue clone @rprichard's [wslbridge] repository.

    - Apply the following patch which _replaces_ `wsl.exe` with `bash.exe`
      and `getpwuid()` with `getenv("SHELL")` for static linking.
    - Compile both backend and frontend.
    - Then check if the default distribution is in WSL V1.
    - Run `./out/wslbridge.exe`.
    - Run any command that produces big output. For example, `ls /usr/bin`.

  - Patch file: [wslbridge-conection-broken.patch.txt](https://github.com/Biswa96/wslbridge2/files/3698266/wslbridge-conection-broken.patch.txt)

[wslbridge]: https://github.com/rprichard/wslbridge

**Comment by dxhisboy** [comment]
I used gdb to attach to the backend, it was terminated by a `SIGSEGV`:
```
xduan@xduan-ux410uqk:~$ sudo gdb -p 646
GNU gdb (Ubuntu 8.1-0ubuntu3.1) 8.1.0.20180409-git
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "x86_64-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word".
Attaching to process 646
[New LWP 648]
[New LWP 649]
[New LWP 650]
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
0x000000000042240a in __waitpid (pid=pid@entry=647, stat_loc=stat_loc@entry=0x7ffffc516cd0, options=options@entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:30
30      ../sysdeps/unix/sysv/linux/waitpid.c: No such file or directory.
(gdb) c
Continuing.

Thread 4 "wslbridge-backe" received signal SIGSEGV, Segmentation fault.
[Switching to Thread 0x7fe391d10700 (LWP 650)]
0x0000000000000000 in ?? ()
(gdb) bt
#0  0x0000000000000000 in ?? ()
#1  0x0000000000404fe9 in std::condition_variable::notify_one() ()
#2  0x00000000004028ad in (anonymous namespace)::handlePacket (p=..., ioloop=0x7ffffc516f10) at wslbridge-backend.cc:371
#3  readControlSocketThread<(anonymous namespace)::IoLoop, (anonymous namespace)::handlePacket, (anonymous namespace)::connectionBrokenAbort> (controlSocketFd=3, userObj=0x7ffffc516f10) at ../common/SocketIo.h:153
#4  0x0000000000403ecf in execute_native_thread_routine ()
#5  0x000000000041ed7b in start_thread (arg=0x7fe391d10700) at pthread_create.c:463
#6  0x00000000004732df in clone ()
(gdb)
```
Maybe I need to view the code somehow.

**Comment by Biswa96** [comment]
Thanks for the debugging info.

Some thoughts:

1. I tried to replace the wslbridge logic (std::thread) with the hvpty logic (poll) and it works. So I am thinking to make wslbridge from scratch. But feared to introduce any bug... 😌
2. Is this glibc issue? Because it works with musl.
3. Who breaks the connection first -- frontend or backened?

**Comment by dxhisboy** [comment]
A workaround is to use semaphore in `semaphore.h`, this lies on a much lower level, maybe contains only some system calls.
After using the patch to a fresh `wslbridge` (@Biswa96 's patch is also included in this patch), the problem has gone when I execute `ls /bin/` in an ubuntu WSL (the only distro in WSL1 I have).
[wslbridge-biswa-patch-and-sem.patch.txt](https://github.com/Biswa96/wslbridge2/files/3720210/wslbridge-biswa-patch-and-sem.patch.txt)

**Comment by dxhisboy** [comment]
> Thanks for the debugging info.
> 
> Some thoughts:
> 
> 1. I tried to replace the wslbridge logic (std::thread) with the hvpty logic (poll) and it works. So I am thinking to make wslbridge from scratch. But feared to introduce any bug... 😌
> 2. Is this glibc issue? Because it works with musl.
> 3. Who breaks the connection first -- frontend or backened?

I think it was not a good idea to write this kind of programs in C++. This should be a `libstdc++` issue in my opinion.
If a program needs static linking, we should make sure the statically linked part will not vary too much in different distros.

**Comment by dxhisboy** [comment]
Also, is it possible to record a series operation and replay it to a frontend for testing?

**Comment by Biswa96** [comment]
> is it possible to record a series operation and replay it to a frontend for testing?

This remind me the [The Two Generals’ Problem][1].

[1]: https://www.youtube.com/watch?v=IP-rGJKSZ3s

**Comment by Biswa96** [comment]
I have tried your patch using semaphore but output got stuck in middle. Try with a heavy output like `ls -la /usr/bin | lolcat`.

**Comment by dxhisboy** [comment]
Maybe my ubuntu has too few things in `/bin`, but I reproduced the bug using `xxd /dev/random`.
Also, I see the `xxd` died while I do not know why and `ps -ef` only shows the `wslbridge-backend`.

**Comment by Biswa96** [comment]
I will remove the old logic in wslbridge2 and add the hvpty logic.

**Comment by dxhisboy** [comment]
> I will remove the old logic in wslbridge2 and add the hvpty logic.

I give up debugging this. One bug I found is that the semaphore may overflow when a process generates a lot of outputs. Then I added a loop to prevent the semaphore from overflowing. Then the `childToSocketThread` exits without report an error, then I added an `abort()` at the end of this function, and the thread does call the `abort()`. That is all what I found.
